File: services/google/binaryauthorization/attestor.go
// Copyright 2021 Google LLC. All Rights Reserved.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package binaryauthorization

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"time"

	"google.golang.org/api/googleapi"
	"github.com/GoogleCloudPlatform/declarative-resource-client-library/dcl"
)

type Attestor struct {
	Name                 *string                       `json:"name"`
	Description          *string                       `json:"description"`
	UserOwnedDrydockNote *AttestorUserOwnedDrydockNote `json:"userOwnedDrydockNote"`
	UpdateTime           *string                       `json:"updateTime"`
	Project              *string                       `json:"project"`
}

func (r *Attestor) String() string {
	return dcl.SprintResource(r)
}

// The enum AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum.
type AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum string

// AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnumRef returns a *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum with the value of string s
// If the empty string is provided, nil is returned.
func AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnumRef(s string) *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum {
	if s == "" {
		return nil
	}

	v := AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum(s)
	return &v
}

func (v AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum) Validate() error {
	for _, s := range []string{"SIGNATURE_ALGORITHM_UNSPECIFIED", "RSA_PSS_2048_SHA256", "RSA_PSS_3072_SHA256", "RSA_PSS_4096_SHA256", "RSA_PSS_4096_SHA512", "RSA_SIGN_PKCS1_2048_SHA256", "RSA_SIGN_PKCS1_3072_SHA256", "RSA_SIGN_PKCS1_4096_SHA256", "RSA_SIGN_PKCS1_4096_SHA512", "ECDSA_P256_SHA256", "EC_SIGN_P256_SHA256", "ECDSA_P384_SHA384", "EC_SIGN_P384_SHA384", "ECDSA_P521_SHA512", "EC_SIGN_P521_SHA512"} {
		if string(v) == s {
			return nil
		}
	}
	return &dcl.EnumInvalidError{
		Enum:  "AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum",
		Value: string(v),
		Valid: []string{},
	}
}

type AttestorUserOwnedDrydockNote struct {
	empty                         bool                                     `json:"-"`
	NoteReference                 *string                                  `json:"noteReference"`
	PublicKeys                    []AttestorUserOwnedDrydockNotePublicKeys `json:"publicKeys"`
	DelegationServiceAccountEmail *string                                  `json:"delegationServiceAccountEmail"`
}

type jsonAttestorUserOwnedDrydockNote AttestorUserOwnedDrydockNote

func (r *AttestorUserOwnedDrydockNote) UnmarshalJSON(data []byte) error {
	var res jsonAttestorUserOwnedDrydockNote
	if err := json.Unmarshal(data, &res); err != nil {
		return err
	}

	var m map[string]interface{}
	json.Unmarshal(data, &m)

	if len(m) == 0 {
		*r = *EmptyAttestorUserOwnedDrydockNote
	} else {

		r.NoteReference = res.NoteReference

		r.PublicKeys = res.PublicKeys

		r.DelegationServiceAccountEmail = res.DelegationServiceAccountEmail

	}
	return nil
}

// This object is used to assert a desired state where this AttestorUserOwnedDrydockNote is
// empty.  Go lacks global const objects, but this object should be treated
// as one.  Modifying this object will have undesirable results.
var EmptyAttestorUserOwnedDrydockNote *AttestorUserOwnedDrydockNote = &AttestorUserOwnedDrydockNote{empty: true}

func (r *AttestorUserOwnedDrydockNote) Empty() bool {
	return r.empty
}

func (r *AttestorUserOwnedDrydockNote) String() string {
	return dcl.SprintResource(r)
}

func (r *AttestorUserOwnedDrydockNote) HashCode() string {
	// Placeholder for a more complex hash method that handles ordering, etc
	// Hash resource body for easy comparison later
	hash := sha256.New().Sum([]byte(r.String()))
	return fmt.Sprintf("%x", hash)
}

type AttestorUserOwnedDrydockNotePublicKeys struct {
	empty                    bool                                                 `json:"-"`
	Comment                  *string                                              `json:"comment"`
	Id                       *string                                              `json:"id"`
	AsciiArmoredPgpPublicKey *string                                              `json:"asciiArmoredPgpPublicKey"`
	PkixPublicKey            *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey `json:"pkixPublicKey"`
}

type jsonAttestorUserOwnedDrydockNotePublicKeys AttestorUserOwnedDrydockNotePublicKeys

func (r *AttestorUserOwnedDrydockNotePublicKeys) UnmarshalJSON(data []byte) error {
	var res jsonAttestorUserOwnedDrydockNotePublicKeys
	if err := json.Unmarshal(data, &res); err != nil {
		return err
	}

	var m map[string]interface{}
	json.Unmarshal(data, &m)

	if len(m) == 0 {
		*r = *EmptyAttestorUserOwnedDrydockNotePublicKeys
	} else {

		r.Comment = res.Comment

		r.Id = res.Id

		r.AsciiArmoredPgpPublicKey = res.AsciiArmoredPgpPublicKey

		r.PkixPublicKey = res.PkixPublicKey

	}
	return nil
}

// This object is used to assert a desired state where this AttestorUserOwnedDrydockNotePublicKeys is
// empty.  Go lacks global const objects, but this object should be treated
// as one.  Modifying this object will have undesirable results.
var EmptyAttestorUserOwnedDrydockNotePublicKeys *AttestorUserOwnedDrydockNotePublicKeys = &AttestorUserOwnedDrydockNotePublicKeys{empty: true}

func (r *AttestorUserOwnedDrydockNotePublicKeys) Empty() bool {
	return r.empty
}

func (r *AttestorUserOwnedDrydockNotePublicKeys) String() string {
	return dcl.SprintResource(r)
}

func (r *AttestorUserOwnedDrydockNotePublicKeys) HashCode() string {
	// Placeholder for a more complex hash method that handles ordering, etc
	// Hash resource body for easy comparison later
	hash := sha256.New().Sum([]byte(r.String()))
	return fmt.Sprintf("%x", hash)
}

type AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey struct {
	empty              bool                                                                       `json:"-"`
	PublicKeyPem       *string                                                                    `json:"publicKeyPem"`
	SignatureAlgorithm *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKeySignatureAlgorithmEnum `json:"signatureAlgorithm"`
}

type jsonAttestorUserOwnedDrydockNotePublicKeysPkixPublicKey AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey

func (r *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey) UnmarshalJSON(data []byte) error {
	var res jsonAttestorUserOwnedDrydockNotePublicKeysPkixPublicKey
	if err := json.Unmarshal(data, &res); err != nil {
		return err
	}

	var m map[string]interface{}
	json.Unmarshal(data, &m)

	if len(m) == 0 {
		*r = *EmptyAttestorUserOwnedDrydockNotePublicKeysPkixPublicKey
	} else {

		r.PublicKeyPem = res.PublicKeyPem

		r.SignatureAlgorithm = res.SignatureAlgorithm

	}
	return nil
}

// This object is used to assert a desired state where this AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey is
// empty.  Go lacks global const objects, but this object should be treated
// as one.  Modifying this object will have undesirable results.
var EmptyAttestorUserOwnedDrydockNotePublicKeysPkixPublicKey *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey = &AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey{empty: true}

func (r *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey) Empty() bool {
	return r.empty
}

func (r *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey) String() string {
	return dcl.SprintResource(r)
}

func (r *AttestorUserOwnedDrydockNotePublicKeysPkixPublicKey) HashCode() string {
	// Placeholder for a more complex hash method that handles ordering, etc
	// Hash resource body for easy comparison later
	hash := sha256.New().Sum([]byte(r.String()))
	return fmt.Sprintf("%x", hash)
}

// Describe returns a simple description of this resource to ensure that automated tools
// can identify it.
func (r *Attestor) Describe() dcl.ServiceTypeVersion {
	return dcl.ServiceTypeVersion{
		Service: "binary_authorization",
		Type:    "Attestor",
		Version: "binaryauthorization",
	}
}

const AttestorMaxPage = -1

type AttestorList struct {
	Items []*Attestor

	nextToken string

	pageSize int32

	resource *Attestor
}

func (l *AttestorList) HasNext() bool {
	return l.nextToken != ""
}

func (l *AttestorList) Next(ctx context.Context, c *Client) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.TimeoutOr(0*time.Second))
	defer cancel()

	if !l.HasNext() {
		return fmt.Errorf("no next page")
	}
	items, token, err := c.listAttestor(ctx, l.resource, l.nextToken, l.pageSize)
	if err != nil {
		return err
	}
	l.Items = items
	l.nextToken = token
	return err
}

func (c *Client) ListAttestor(ctx context.Context, r *Attestor) (*AttestorList, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.TimeoutOr(0*time.Second))
	defer cancel()

	return c.ListAttestorWithMaxResults(ctx, r, AttestorMaxPage)

}

func (c *Client) ListAttestorWithMaxResults(ctx context.Context, r *Attestor, pageSize int32) (*AttestorList, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.TimeoutOr(0*time.Second))
	defer cancel()

	items, token, err := c.listAttestor(ctx, r, "", pageSize)
	if err != nil {
		return nil, err
	}
	return &AttestorList{
		Items:     items,
		nextToken: token,
		pageSize:  pageSize,
		resource:  r,
	}, nil
}

func (c *Client) GetAttestor(ctx context.Context, r *Attestor) (*Attestor, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.TimeoutOr(0*time.Second))
	defer cancel()

	b, err := c.getAttestorRaw(ctx, r)
	if err != nil {
		if dcl.IsNotFound(err) {
			return nil, &googleapi.Error{
				Code:    404,
				Message: err.Error(),
			}
		}
		return nil, err
	}
	result, err := unmarshalAttestor(b, c)
	if err != nil {
		return nil, err
	}
	result.Project = r.Project
	result.Name = r.Name

	c.Config.Logger.Infof("Retrieved raw result state: %v", result)
	c.Config.Logger.Infof("Canonicalizing with specified state: %v", r)
	result, err = canonicalizeAttestorNewState(c, result, r)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Created result state: %v", result)

	return result, nil
}

func (c *Client) DeleteAttestor(ctx context.Context, r *Attestor) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.TimeoutOr(0*time.Second))
	defer cancel()

	if r == nil {
		return fmt.Errorf("Attestor resource is nil")
	}
	c.Config.Logger.Info("Deleting Attestor...")
	deleteOp := deleteAttestorOperation{}
	return deleteOp.do(ctx, r, c)
}

// DeleteAllAttestor deletes all resources that the filter functions returns true on.
func (c *Client) DeleteAllAttestor(ctx context.Context, r *Attestor, filter func(*Attestor) bool) error {
	listObj, err := c.ListAttestor(ctx, r)
	if err != nil {
		return err
	}

	err = c.deleteAllAttestor(ctx, filter, listObj.Items)
	if err != nil {
		return err
	}
	for listObj.HasNext() {
		err = listObj.Next(ctx, c)
		if err != nil {
			return nil
		}
		err = c.deleteAllAttestor(ctx, filter, listObj.Items)
		if err != nil {
			return err
		}
	}
	return nil
}

func (c *Client) ApplyAttestor(ctx context.Context, rawDesired *Attestor, opts ...dcl.ApplyOption) (*Attestor, error) {
	var resultNewState *Attestor
	err := dcl.Do(ctx, func(ctx context.Context) (*dcl.RetryDetails, error) {
		newState, err := applyAttestorHelper(c, ctx, rawDesired, opts...)
		resultNewState = newState
		if err != nil {
			// If the error is 409, there is conflict in resource update.
			// Here we want to apply changes based on latest state.
			if dcl.IsConflictError(err) {
				return &dcl.RetryDetails{}, dcl.OperationNotDone{Err: err}
			}
			return nil, err
		}
		return nil, nil
	}, c.Config.RetryProvider)
	return resultNewState, err
}

func applyAttestorHelper(c *Client, ctx context.Context, rawDesired *Attestor, opts ...dcl.ApplyOption) (*Attestor, error) {
	c.Config.Logger.Info("Beginning ApplyAttestor...")
	c.Config.Logger.Infof("User specified desired state: %v", rawDesired)

	ctx, cancel := context.WithTimeout(ctx, c.Config.TimeoutOr(0*time.Second))
	defer cancel()

	// 1.1: Validation of user-specified fields in desired state.
	if err := rawDesired.validate(); err != nil {
		return nil, err
	}

	initial, desired, fieldDiffs, err := c.attestorDiffsForRawDesired(ctx, rawDesired, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to create a diff: %w", err)
	}

	diffs, err := convertFieldDiffsToAttestorDiffs(c.Config, fieldDiffs, opts)
	if err != nil {
		return nil, err
	}

	// TODO(magic-modules-eng): 2.2 Feasibility check (all updates are feasible so far).

	// 2.3: Lifecycle Directive Check
	var create bool
	var recreate bool
	lp := dcl.FetchLifecycleParams(opts)
	if initial == nil {
		if dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
			return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Creation blocked by lifecycle params: %#v.", desired)}
		}
		create = true
	} else if dcl.HasLifecycleParam(lp, dcl.BlockAcquire) {
		return nil, dcl.ApplyInfeasibleError{
			Message: fmt.Sprintf("Resource already exists - apply blocked by lifecycle params: %#v.", initial),
		}
	} else {
		for _, d := range diffs {
			if d.RequiresRecreate {
				if dcl.HasLifecycleParam(lp, dcl.BlockDestruction) || dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
					return nil, dcl.ApplyInfeasibleError{
						Message: fmt.Sprintf("Infeasible update: (%v) would require recreation.", d),
					}
				}
				recreate = true
			}
			if dcl.HasLifecycleParam(lp, dcl.BlockModification) {
				return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Modification blocked, diff (%v) unresolvable.", d)}
			}
		}
	}

	// 2.4 Imperative Request Planning
	var ops []attestorApiOperation
	if create {
		ops = append(ops, &createAttestorOperation{})
	} else if recreate {
		ops = append(ops, &deleteAttestorOperation{})
		ops = append(ops, &createAttestorOperation{})
		// We should re-canonicalize based on a nil existing resource.
		desired, err = canonicalizeAttestorDesiredState(rawDesired, nil)
		if err != nil {
			return nil, err
		}
	} else {
		for _, d := range diffs {
			ops = append(ops, d.UpdateOp)
		}
	}
	c.Config.Logger.Infof("Created plan: %#v", ops)

	// 2.5 Request Actuation
	for _, op := range ops {
		c.Config.Logger.Infof("Performing operation %#v", op)
		if err := op.do(ctx, desired, c); err != nil {
			c.Config.Logger.Infof("Failed operation %#v: %v", op, err)
			return nil, err
		}
		c.Config.Logger.Infof("Finished operation %#v", op)
	}

	// 3.1, 3.2a Retrieval of raw new state & canonicalization with desired state
	c.Config.Logger.Info("Retrieving raw new state...")
	rawNew, err := c.GetAttestor(ctx, desired.urlNormalized())
	if err != nil {
		return nil, err
	}
	// Get additional values from the first response.
	// These values should be merged into the newState above.
	if len(ops) > 0 {
		lastOp := ops[len(ops)-1]
		if o, ok := lastOp.(*createAttestorOperation); ok {
			if r, hasR := o.FirstResponse(); hasR {

				c.Config.Logger.Info("Retrieving raw new state from operation...")

				fullResp, err := unmarshalMapAttestor(r, c)
				if err != nil {
					return nil, err
				}

				rawNew, err = canonicalizeAttestorNewState(c, rawNew, fullResp)
				if err != nil {
					return nil, err
				}
			}
		}
	}

	c.Config.Logger.Infof("Canonicalizing with raw desired state: %v", rawDesired)
	// 3.2b Canonicalization of raw new state using raw desired state
	newState, err := canonicalizeAttestorNewState(c, rawNew, rawDesired)
	if err != nil {
		return nil, err
	}

	c.Config.Logger.Infof("Created canonical new state: %v", newState)
	// 3.3 Comparison of the new state and raw desired state.
	// TODO(magic-modules-eng): EVENTUALLY_CONSISTENT_UPDATE
	newDesired, err := canonicalizeAttestorDesiredState(rawDesired, newState)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Diffing using canonicalized desired state: %v", newDesired)
	newDiffs, err := diffAttestor(c, newDesired, newState)
	if err != nil {
		return nil, err
	}

	if len(newDiffs) == 0 {
		c.Config.Logger.Info("No diffs found. Apply was successful.")
	} else {
		c.Config.Logger.Infof("Found diffs: %v", newDiffs)
		diffMessages := make([]string, len(newDiffs))
		for i, d := range newDiffs {
			diffMessages[i] = fmt.Sprintf("%v", d)
		}
		return newState, dcl.DiffAfterApplyError{Diffs: diffMessages}
	}
	c.Config.Logger.Info("Done Apply.")
	return newState, nil
}

File: python/services/binaryauthorization/attestor.py
# Copyright 2021 Google LLC. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from connector import channel
from google3.cloud.graphite.mmv2.services.google.binary_authorization import (
    attestor_pb2,
)
from google3.cloud.graphite.mmv2.services.google.binary_authorization import (
    attestor_pb2_grpc,
)

from typing import List


class Attestor(object):
    def __init__(
        self,
        name: str = None,
        description: str = None,
        user_owned_grafeas_note: dict = None,
        update_time: str = None,
        project: str = None,
        service_account_file: str = "",
    ):

        channel.initialize()
        self.name = name
        self.description = description
        self.user_owned_grafeas_note = user_owned_grafeas_note
        self.project = project
        self.service_account_file = service_account_file

    def apply(self):
        stub = attestor_pb2_grpc.BinaryauthorizationAttestorServiceStub(
            channel.Channel()
        )
        request = attestor_pb2.ApplyBinaryauthorizationAttestorRequest()
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.description):
            request.resource.description = Primitive.to_proto(self.description)

        if AttestorUserOwnedGrafeasNote.to_proto(self.user_owned_grafeas_note):
            request.resource.user_owned_grafeas_note.CopyFrom(
                AttestorUserOwnedGrafeasNote.to_proto(self.user_owned_grafeas_note)
            )
        else:
            request.resource.ClearField("user_owned_grafeas_note")
        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        request.service_account_file = self.service_account_file

        response = stub.ApplyBinaryauthorizationAttestor(request)
        self.name = Primitive.from_proto(response.name)
        self.description = Primitive.from_proto(response.description)
        self.user_owned_grafeas_note = AttestorUserOwnedGrafeasNote.from_proto(
            response.user_owned_grafeas_note
        )
        self.update_time = Primitive.from_proto(response.update_time)
        self.project = Primitive.from_proto(response.project)

    def delete(self):
        stub = attestor_pb2_grpc.BinaryauthorizationAttestorServiceStub(
            channel.Channel()
        )
        request = attestor_pb2.DeleteBinaryauthorizationAttestorRequest()
        request.service_account_file = self.service_account_file
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.description):
            request.resource.description = Primitive.to_proto(self.description)

        if AttestorUserOwnedGrafeasNote.to_proto(self.user_owned_grafeas_note):
            request.resource.user_owned_grafeas_note.CopyFrom(
                AttestorUserOwnedGrafeasNote.to_proto(self.user_owned_grafeas_note)
            )
        else:
            request.resource.ClearField("user_owned_grafeas_note")
        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        response = stub.DeleteBinaryauthorizationAttestor(request)

    @classmethod
    def list(self, project, service_account_file=""):
        stub = attestor_pb2_grpc.BinaryauthorizationAttestorServiceStub(
            channel.Channel()
        )
        request = attestor_pb2.ListBinaryauthorizationAttestorRequest()
        request.service_account_file = service_account_file
        request.Project = project

        return stub.ListBinaryauthorizationAttestor(request).items

    def to_proto(self):
        resource = attestor_pb2.BinaryauthorizationAttestor()
        if Primitive.to_proto(self.name):
            resource.name = Primitive.to_proto(self.name)
        if Primitive.to_proto(self.description):
            resource.description = Primitive.to_proto(self.description)
        if AttestorUserOwnedGrafeasNote.to_proto(self.user_owned_grafeas_note):
            resource.user_owned_grafeas_note.CopyFrom(
                AttestorUserOwnedGrafeasNote.to_proto(self.user_owned_grafeas_note)
            )
        else:
            resource.ClearField("user_owned_grafeas_note")
        if Primitive.to_proto(self.project):
            resource.project = Primitive.to_proto(self.project)
        return resource


class AttestorUserOwnedGrafeasNote(object):
    def __init__(
        self,
        note_reference: str = None,
        public_keys: list = None,
        delegation_service_account_email: str = None,
    ):
        self.note_reference = note_reference
        self.public_keys = public_keys
        self.delegation_service_account_email = delegation_service_account_email

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = attestor_pb2.BinaryauthorizationAttestorUserOwnedGrafeasNote()
        if Primitive.to_proto(resource.note_reference):
            res.note_reference = Primitive.to_proto(resource.note_reference)
        if AttestorUserOwnedGrafeasNotePublicKeysArray.to_proto(resource.public_keys):
            res.public_keys.extend(
                AttestorUserOwnedGrafeasNotePublicKeysArray.to_proto(
                    resource.public_keys
                )
            )
        if Primitive.to_proto(resource.delegation_service_account_email):
            res.delegation_service_account_email = Primitive.to_proto(
                resource.delegation_service_account_email
            )
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return AttestorUserOwnedGrafeasNote(
            note_reference=Primitive.from_proto(resource.note_reference),
            public_keys=AttestorUserOwnedGrafeasNotePublicKeysArray.from_proto(
                resource.public_keys
            ),
            delegation_service_account_email=Primitive.from_proto(
                resource.delegation_service_account_email
            ),
        )


class AttestorUserOwnedGrafeasNoteArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [AttestorUserOwnedGrafeasNote.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [AttestorUserOwnedGrafeasNote.from_proto(i) for i in resources]


class AttestorUserOwnedGrafeasNotePublicKeys(object):
    def __init__(
        self,
        comment: str = None,
        id: str = None,
        ascii_armored_pgp_public_key: str = None,
        pkix_public_key: dict = None,
    ):
        self.comment = comment
        self.id = id
        self.ascii_armored_pgp_public_key = ascii_armored_pgp_public_key
        self.pkix_public_key = pkix_public_key

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = attestor_pb2.BinaryauthorizationAttestorUserOwnedGrafeasNotePublicKeys()
        if Primitive.to_proto(resource.comment):
            res.comment = Primitive.to_proto(resource.comment)
        if Primitive.to_proto(resource.id):
            res.id = Primitive.to_proto(resource.id)
        if Primitive.to_proto(resource.ascii_armored_pgp_public_key):
            res.ascii_armored_pgp_public_key = Primitive.to_proto(
                resource.ascii_armored_pgp_public_key
            )
        if AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey.to_proto(
            resource.pkix_public_key
        ):
            res.pkix_public_key.CopyFrom(
                AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey.to_proto(
                    resource.pkix_public_key
                )
            )
        else:
            res.ClearField("pkix_public_key")
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return AttestorUserOwnedGrafeasNotePublicKeys(
            comment=Primitive.from_proto(resource.comment),
            id=Primitive.from_proto(resource.id),
            ascii_armored_pgp_public_key=Primitive.from_proto(
                resource.ascii_armored_pgp_public_key
            ),
            pkix_public_key=AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey.from_proto(
                resource.pkix_public_key
            ),
        )


class AttestorUserOwnedGrafeasNotePublicKeysArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [AttestorUserOwnedGrafeasNotePublicKeys.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [AttestorUserOwnedGrafeasNotePublicKeys.from_proto(i) for i in resources]


class AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey(object):
    def __init__(self, public_key_pem: str = None, signature_algorithm: str = None):
        self.public_key_pem = public_key_pem
        self.signature_algorithm = signature_algorithm

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = (
            attestor_pb2.BinaryauthorizationAttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey()
        )
        if Primitive.to_proto(resource.public_key_pem):
            res.public_key_pem = Primitive.to_proto(resource.public_key_pem)
        if AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum.to_proto(
            resource.signature_algorithm
        ):
            res.signature_algorithm = AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum.to_proto(
                resource.signature_algorithm
            )
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey(
            public_key_pem=Primitive.from_proto(resource.public_key_pem),
            signature_algorithm=AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum.from_proto(
                resource.signature_algorithm
            ),
        )


class AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeyArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [
            AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey.to_proto(i)
            for i in resources
        ]

    @classmethod
    def from_proto(self, resources):
        return [
            AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKey.from_proto(i)
            for i in resources
        ]


class AttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum(object):
    @classmethod
    def to_proto(self, resource):
        if not resource:
            return resource
        return attestor_pb2.BinaryauthorizationAttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum.Value(
            "BinaryauthorizationAttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum%s"
            % resource
        )

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return resource
        return attestor_pb2.BinaryauthorizationAttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum.Name(
            resource
        )[
            len(
                "BinaryauthorizationAttestorUserOwnedGrafeasNotePublicKeysPkixPublicKeySignatureAlgorithmEnum"
            ) :
        ]


class Primitive(object):
    @classmethod
    def to_proto(self, s):
        if not s:
            return ""
        return s

    @classmethod
    def from_proto(self, s):
        return s
